binary_search_tree/tree.py

```diff
@@ -9,50 +13,144 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: 0(log n)
+    # Space Complexity: 0(log n) 
+    def add_helper(self, current_node, key, value):
+        # helper takes advantage of recursive
+        # if node is empty, return new node with key and value
+        if current_node == None:
+            return TreeNode(key, value)
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value) 
+        return current_node
+            
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+            
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: 0(log n)
+    # Space Complexity: 0(log n)
```

Comment:
Luckily since you did an iterative approach for this, we do not take the hit of the space that will be added to the stack if you had done this recursively. As such, the space complexity is O(1).

Reply:
Ignore my comment on how there is space complexity savings for a recursive solution. Both should be O(1) in my opinion because they are not using additional space. I was going off the instructor solution and then realized that I think they made a mistake.